android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/RoutingSource.kt

```diff
@@ -0,0 +1,190 @@
+
+        companion object {
+            fun <C : Parcelable> permanent(permanents: Iterable<C>) = Permanent(permanents)
+
+            fun <C : Parcelable> permanent(vararg permanents: C) = Permanent(permanents.toSet())
+        }
+
+        private val routingElements =
+            permanents.mapIndexed { idx, configuration ->
+                RoutingHistoryElement(
+                    activation = ACTIVE,
+                    routing = Routing(
+                        configuration = configuration,
+                        identifier = Identifier("Permanent $idx")
+                    )
+            ) }
+
+        private val permanentHistory =
+            RoutingHistory.from(routingElements)
+
+        override fun remove(identifier: Identifier) {
+            // no-op -- it's permanent!
+        }
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) {
+            Observable
+                .just(permanentHistory)
+                .subscribe(observer)
+        }
+    }
+
+    // TODO extract
+    class Empty<C : Parcelable> : RoutingSource<C> {
+        override fun remove(identifier: Identifier) {
+            // no-op -- it's empty
+        }
+
+        override fun subscribe(observer: Observer<in RoutingHistory<C>>) =
+            Observable.empty<RoutingHistory<C>>().subscribe(observer)
+    }
+
+    // TODO extract
+    class Pool<C : Parcelable>(
+        private val allowRepeatingConfigurations: Boolean
+    ) : RoutingSource<C> {
+        private var elements: Map<Identifier, RoutingHistoryElement<C>> = emptyMap()
+        private val current: RoutingHistory<C>
+            get() = RoutingHistory.from(elements.values)
+
+        private val states: Relay<RoutingHistory<C>> = BehaviorRelay.createDefault(
+            RoutingHistory.from(current)
+        )
+
+        fun add(
+            configuration: C,
+            identifier: Identifier = Identifier(
+                "Set ${System.identityHashCode(this)} #$configuration"
+            )
+        ): Identifier {
+            if (!allowRepeatingConfigurations) {
+                elements
+                    .filter { it.value.routing.configuration == configuration }
+                    .map { it.key }
+                    .firstOrNull()?.let {
+                        return it
+                    }
+            }
+
+            elements = elements.plus(
+                identifier to RoutingHistoryElement(
+                    activation = INACTIVE,
+                    routing = Routing(
+                        configuration = configuration,
+                        identifier = identifier
+                    ),
+                    // TODO consider overlay support -- not needed now, can be added later
+                    overlays = emptyList()
+                )
+            )
+
+            updateState()
+
+            return identifier
+        }
+
+        fun activate(identifier: Identifier) {
```

Comment:
will it be useful information if there is no such identifier? May be such methods should return `boolean` or `Identifier?`?